core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala

```diff
@@ -48,32 +56,74 @@ private[spark] class DiskBlockManager(conf: SparkConf, deleteFilesOnStop: Boolea
   // of subDirs(i) is protected by the lock of subDirs(i)
   private val subDirs = Array.fill(localDirs.length)(new Array[File](subDirsPerLocalDir))
 
+  private[spark] val badDirs = ArrayBuffer[File]()
+  private[spark] val dirToBlacklistExpiryTime = new HashMap[File, Long]
+  // Filename to dirId, it should be small enough to put into memory
+  private[spark] val migratedDirIdIndex = new ConcurrentHashMap[String, Int].asScala
+
   private val shutdownHook = addShutdownHook()
 
   /** Looks up a file by hashing it into one of our local subdirectories. */
   // This method should be kept in sync with
   // org.apache.spark.network.shuffle.ExternalShuffleBlockResolver#getFile().
   def getFile(filename: String): File = {
+    var mostRecentFailure: Exception = null
     // Figure out which local directory it hashes to, and which subdirectory in that
     val hash = Utils.nonNegativeHash(filename)
-    val dirId = hash % localDirs.length
+    val dirId = migratedDirIdIndex.getOrElse(filename, hash % localDirs.length)
     val subDirId = (hash / localDirs.length) % subDirsPerLocalDir
 
     // Create the subdirectory if it doesn't already exist
     val subDir = subDirs(dirId).synchronized {
+      // Update blacklist
+      val now = clock.getTimeMillis()
+      badDirs.synchronized {
+        val unblacklisted = badDirs.filter(now >= dirToBlacklistExpiryTime(_))
+        unblacklisted.foreach { dir =>
+          badDirs -= dir
+          dirToBlacklistExpiryTime.remove(dir)
+        }
+      }
+
       val old = subDirs(dirId)(subDirId)
       if (old != null) {
         old
       } else {
-        val newDir = new File(localDirs(dirId), "%02x".format(subDirId))
-        if (!newDir.exists() && !newDir.mkdir()) {
-          throw new IOException(s"Failed to create local dir in $newDir.")
+        assert(!migratedDirIdIndex.contains(filename))
+        var newDir: File = null
+        for (attempt <- 0 until maxAttempts if newDir == null) {
+          val goodDirId = badDirs.synchronized {
+            val isBlacklisted = badDirs.contains(localDirs(dirId))
+            if (isBlacklisted) {
+              localDirs.indexWhere(!badDirs.contains(_))
+            } else {
+              dirId
+            }
+          }
+          try {
+            if (goodDirId < 0) {
+              throw new IOException("No good disk directories available")
```

Comment:
Perhaps say all disks blacklisted

Reply:
Oh I see error is printed below,  the message below probably makes more sense to print number  of disks blacklisted.

Reply:
done